sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -0,0 +1,388 @@
+      // time, so we process those stages too in order to reduce re-planning.
+      val nextMsg = events.take()
+      val rem = new util.ArrayList[StageMaterializationEvent]()
+      events.drainTo(rem)
+      (Seq(nextMsg) ++ rem.asScala).foreach {
+        case StageSuccess(stage, res) =>
+          stage.resultOption = Some(res)
+          completedStages += stage.id
+        case StageFailure(stage, ex) =>
+          errors.append(
+            new SparkException(s"Fail to materialize query stage: ${stage.treeString}", ex))
+      }
+
+      // In case of errors, we cancel all running stages and throw exception.
+      if (errors.nonEmpty) {
+        try {
+          currentPhysicalPlan.foreach {
+            case s: QueryStageExec => s.cancel()
+            case _ =>
+          }
+        } finally {
+          val ex = new SparkException(
+            "Adaptive execution failed due to stage materialization failures.", errors.head)
+          errors.tail.foreach(ex.addSuppressed)
+          throw ex
+        }
+      }
+
+      // Do re-planning and try creating new stages on the new physical plan.
+      reOptimize()
+      result = createQueryStages(currentPhysicalPlan)
+    }
+
+    // Run the final plan when there's no more unfinished stages.
+    currentPhysicalPlan =
+      AdaptiveSparkPlanExec.applyPhysicalRules(result.newPlan, queryStageOptimizerRules)
+    logDebug(s"Final plan: $currentPhysicalPlan")
+    onUpdatePlan()
+    currentPhysicalPlan.execute()
+  }
+
+  override def generateTreeString(
+    depth: Int,
+    lastChildren: Seq[Boolean],
+    append: String => Unit,
+    verbose: Boolean,
+    prefix: String = "",
+    addSuffix: Boolean = false,
+    maxFields: Int): Unit = {
+    currentPhysicalPlan.generateTreeString(
+      depth, lastChildren, append, verbose, "", false, maxFields)
+  }
+
+  @transient private val executionId = Option(
+    session.sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)).map(_.toLong)
+
+  @transient private val lock = new Object()
+
+  @transient private val completedStages = mutable.HashSet.empty[Int]
+
+  @volatile private var currentStageId = 0
+
+  @volatile private var currentPhysicalPlan = initialPlan
+
+  @volatile private var currentLogicalPlan = initialPlan.logicalLink.get
+
+  // The logical plan optimizer for re-optimizing the current logical plan.
+  private object Optimizer extends RuleExecutor[LogicalPlan] {
+    // TODO add more optimization rules
+    override protected def batches: Seq[Batch] = Seq()
+  }
+
+  // A list of physical plan rules to be applied before creation of query stages. The physical
+  // plan should reach a final status of query stages (i.e., no more addition or removal of
+  // Exchange nodes) after running these rules.
+  @transient private val queryStagePreparationRules: Seq[Rule[SparkPlan]] =
+    AdaptiveSparkPlanExec.createQueryStagePreparationRules(conf, subqueryMap)
```

Comment:
What `conf` are you using here? I think `QueryPlan.conf` could point to a different configuration than the one in the `session` field. We may want to `override def conf` to make sure they are the same.